Classes/Fusion/Eel/FlowQueryOperations/SortRecursiveByIndexOperation.php

```diff
@@ -50,42 +60,78 @@ public function evaluate(FlowQuery $flowQuery, array $arguments)
+            $pathMap[$nodeIdentifier] = array_merge([$node->aggregateId], array_map(static fn($ancestor) => $ancestor->aggregateId, iterator_to_array($ancestors)));
         }
 
         $flip = $sortOrder === 'DESC' ? -1 : 1;
 
-        usort($nodes, function (NodeInterface $a, NodeInterface $b) use ($indexPathCache, $flip) {
-            if ($a === $b) {
+        usort($nodes, function (Node $a, Node $b) use ($subgraph, $pathMap, $flip) {
+            // Both nodes are equal
+            if ($a->equals($b)) {
                 return 0;
             }
 
-            // Compare index path starting from the site root until a difference is found
-            $aIndexPath = $indexPathCache[$a->getIdentifier()];
-            $bIndexPath = $indexPathCache[$b->getIdentifier()];
-            while (count($aIndexPath) > 0 && count($bIndexPath) > 0) {
-                $diff = (array_pop($aIndexPath) - array_pop($bIndexPath));
-                if ($diff !== 0) {
-                    return $flip * $diff < 0 ? -1 : 1;
+            $commonParentPathSegmentNodeAggregateId = null;
+            $childNodesCache = [];
+
+            // Compare path starting from the site root until a difference is found.
+            $aPath = $pathMap[$a->aggregateId->value];
+            $bPath = $pathMap[$b->aggregateId->value];
+            while (count($aPath) > 0 && count($bPath) > 0) {
+
+                /** @var NodeAggregateId $aPathSegmentNodeAggregateId */
+                $aPathSegmentNodeAggregateId = array_pop($aPath);
+                /** @var NodeAggregateId $bPathSegmentNodeAggregateId */
+                $bPathSegmentNodeAggregateId = array_pop($bPath);
+
+                $pathDiff = (!$aPathSegmentNodeAggregateId->equals($bPathSegmentNodeAggregateId));
+
+                if ($pathDiff === true) {
+                    // Path is different at this segment, so we need to figure out their position under the last common parent.
+                    if ($commonParentPathSegmentNodeAggregateId === null) {
+                        return 0;
+                    }
+
+                    if (!isset($childNodesCache[$commonParentPathSegmentNodeAggregateId->value])) {
+                        $childNodesCache[$commonParentPathSegmentNodeAggregateId->value] = $subgraph->findChildNodes($commonParentPathSegmentNodeAggregateId, FindChildNodesFilter::create());
+                    }
+
+                    $positionDiff = $this->getIndexOfNodeAggregateIdInNodes($childNodesCache[$commonParentPathSegmentNodeAggregateId->value], $aPathSegmentNodeAggregateId)
+                        - $this->getIndexOfNodeAggregateIdInNodes($childNodesCache[$commonParentPathSegmentNodeAggregateId->value], $bPathSegmentNodeAggregateId);
+                    return $flip * $positionDiff < 0 ? -1 : 1;
                 }
+                // No diff in path, we need to go deeper, or they are eventually equal
+                $commonParentPathSegmentNodeAggregateId = $aPathSegmentNodeAggregateId;
```

Comment:
Absolutely not your fault, but I find it really hard to review this. And since there are no tests, it's not easy to verify the desired behavior

Reply:
Yes, absolutly. But I can provide that for each package, where I just want to help out to get this package compatible with Neos 9.